### DISCUSSION TITLE
Feature names for forecasters from make_reduction

### DISCUSSION THREAD

**Discussion by GrzegorzRut** [alan-turing-institute/sktime#2672] [Q&A]
Is there a way to find names of features used in forecasters from make_reduction?

I was investigating the order of features in forecasters from `make_reduction`. It's hard to tell what coefficients from `.coef_` correspond to.
```python
import numpy as np
import pandas as pd
from sklearn.linear_model import ElasticNet
from sktime.forecasting.compose import make_reduction

X = pd.DataFrame({"a": np.arange(0, 50, 5), "b": np.arange(-10,0,1)})
y = pd.Series(np.arange(0, 10))
lm = make_reduction(
    ElasticNet(),
    scitype="tabular-regressor",
    window_length=2,
)

lm.fit(y, X)
lm.estimator_.coef_
```
generates `array([0.        , 0.        , 0.09920628, 0.09661613, 0.        ,
       0.        ])` 

After some digging i found `_sliding_window_transform` which  is responsible for generating features for the regressor:
```python
from sktime.forecasting.base import ForecastingHorizon
from sktime.forecasting.compose._reduce import _sliding_window_transform

fh = ForecastingHorizon([1])
_sliding_window_transform(y, 2, fh, X=X)
```
which generates
```
(array([[2.],
        [3.],
        [4.],
        [5.],
        [6.],
        [7.],
        [8.],
        [9.]]),
 array([[  0.,   1.,   0.,   5., -10.,  -9.],
        [  1.,   2.,   5.,  10.,  -9.,  -8.],
        [  2.,   3.,  10.,  15.,  -8.,  -7.],
        [  3.,   4.,  15.,  20.,  -7.,  -6.],
        [  4.,   5.,  20.,  25.,  -6.,  -5.],
        [  5.,   6.,  25.,  30.,  -5.,  -4.],
        [  6.,   7.,  30.,  35.,  -4.,  -3.],
        [  7.,   8.,  35.,  40.,  -3.,  -2.]]))
```
From this i can deduce the order but i wonder if there is a simpler way to do it.

**Comment by fkiraly** [comment]
@GrzegorzRut, thanks for bringing this up.

Indeed, the reason for not having feature names in the  is internal use of numpy, which eats all the feature names.

There is, in fact, already an open issue to implement the `get_fitted_params` for reducers, that would give convenient access to these params, hopefully with feature names, etc:
https://github.com/alan-turing-institute/sktime/issues/1486

Might be as easy as ensuring the internals work for `pd.DataFrame` throughout, and adding some logic for parameter retrieval...

**Reply by fkiraly** [reply]
FYI @asksteph, related question to your #1486. Any added opinions/insight?

**Reply by asksteph** [reply]
The sklearn estimators can output feature names via the `feature_names_in_` attribute as of version 1.0. If sktime uses versions later than that, then that can be used along with the `coef_` attribute to provide the needed info.

**Reply by fkiraly** [reply]
ah, so all we need to do is pass a `pandas.DataFrame` and all shall be fine?

**Reply by asksteph** [reply]
That's what it seems like based on the sklearn docs. It will return the names of features seen during the fit and is defined only when input X is a dataframe that has feature names that are all strings.

**Reply by GrzegorzRut** [reply]
@asksteph, @fkiraly, many thanks for the answer. 

I'm not sure how general this is but I can confirm that `.feature_names_in_` works with `scitype= "time-series-regressor"` (as, most likely, it will first create a dataset via`_sliding_window_transform`, then it will pass it to `TSFreshFeatureExtractor` which will generate a `pd.DataFrame`).
```python
import numpy as np
import pandas as pd
from sklearn.ensemble import RandomForestRegressor
from sklearn.pipeline import make_pipeline
from sktime.datasets import load_airline
from sktime.forecasting.compose import make_reduction
from sktime.transformations.panel.tsfresh import TSFreshFeatureExtractor

y = load_airline()

regressor = make_pipeline(
    TSFreshFeatureExtractor(show_warnings=False, disable_progressbar=True),
    RandomForestRegressor(),
)
forecaster = make_reduction(
    regressor, scitype="time-series-regressor", window_length=24
)
forecaster.fit(y)

pd.DataFrame(
    zip(
        forecaster.estimator_.steps[-1][1].feature_importances_,
        forecaster.estimator_.steps[-1][1].feature_names_in_,
    )
).rename(columns={0: "val", 1: "var"}).sort_values("val", ascending=False)
```

Similarly, `forecaster.estimator_.steps[-1][1].get_params()` would work in this case as well (this one actually works for `table-regressor` as well).


The problem is that `_sliding_window_transform` is written strictly in `numpy`, and, for this reason, will not generate any new names (old ones would be not enough). Looking at the code, [this](https://github.com/alan-turing-institute/sktime/blob/a761368f32b12913f2d7c22d87776e6112cc785d/sktime/forecasting/compose/_reduce.py#L116) section (`transformers` is None), one could generate the column names as follows:

```python
Xt_columns = [
    f"{obj}_lag_{x}"
    for obj in ["target", *X.columns]
    for x in range(window_length, 0, -1)
]
```
this will not make sense for `scitype="time-series-reggressor"` but could work with `table-reggressor`. Perhaps this could be passed further as an attribute to `make_reduce`.

**Comment by fkiraly** [comment]
@GrzegorzRut, I've started a draft rework of the direct reducer here https://github.com/alan-turing-institute/sktime/pull/2833

partly based on work of yours and of @danbartl, would appreciate a look!

A side effect should be that the linear regressor should now correctly display descriptive feature names.
